File: src/C++/Ejercicio9/Ej9Main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <iomanip>
using namespace std;

// Archivo donde se guardan las ventas
const string ARCHIVO = "ventas.txt";

//  Clase Venta
class Venta {
private:
    string producto;
    int    cantidad;
    double precio;
    double total;

public:
    // Constructor por defecto
    Venta() : producto(""), cantidad(0), precio(0.0), total(0.0) {}

    // Constructor con parámetros
    Venta(string _producto, int _cantidad, double _precio) {
        producto = _producto;
        cantidad = _cantidad;
        precio   = _precio;
        calcularTotal();
    }

    // Métodos GET
    string getProducto() const { return producto; }
    int    getCantidad() const { return cantidad;  }
    double getPrecio()   const { return precio;    }
    double getTotal()    const { return total;     }

    // Métodos SET
    void setProducto(string _producto) { producto = _producto; }
    void setCantidad(int    _cantidad) { cantidad = _cantidad; }
    void setPrecio(double   _precio)   { precio   = _precio;   }

    // Calcular total de la venta
    void calcularTotal() {
        total = cantidad * precio;
    }

    // Mostrar factura de la venta
    void mostrarFactura() const {
        cout << fixed << setprecision(2);
        cout << "  ----------------------------------------" << endl;
        cout << "  Producto  : " << producto           << endl;
        cout << "  Cantidad  : " << cantidad            << endl;
        cout << "  Precio u. : $" << precio             << endl;
        cout << "  TOTAL     : $" << total              << endl;
        cout << "  ----------------------------------------" << endl;
    }
};

//  Función: registrarVenta
//  Solicita datos y guarda la venta en el archivo
void registrarVenta() {
    string producto;
    int    cantidad;
    double precio;

    cout << "\n  -- Registrar Venta --" << endl;
    cout << "  Producto  : "; cin >> producto;

    do {
        cout << "  Cantidad  : "; cin >> cantidad;
        if (cantidad <= 0)
            cout << "  ERROR La cantidad debe ser mayor a 0." << endl;
    } while (cantidad <= 0);

    do {
        cout << "  Precio u. : "; cin >> precio;
        if (precio <= 0)
            cout << "  ERROR El precio debe ser mayor a 0." << endl;
    } while (precio <= 0);

    Venta v(producto, cantidad, precio);

    // Mostrar la factura generada
    cout << "\n  ======== FACTURA GENERADA ========" << endl;
    v.mostrarFactura();

    // Guardar en archivo (modo append)
    ofstream archivo(ARCHIVO, ios::app);
    if (!archivo.is_open()) {
        cout << "  ERROR No se pudo abrir el archivo de ventas." << endl;
        return;
    }

    archivo << fixed << setprecision(2);
    archivo << v.getProducto() << "|"
            << v.getCantidad() << "|"
            << v.getPrecio()   << "|"
            << v.getTotal()    << endl;

    archivo.close();
    cout << " Venta registrada correctamente." << endl;
}

//  Función: mostrarHistorial
//  Lee el archivo y muestra todas las ventas registradas
void mostrarHistorial() {
    ifstream archivo(ARCHIVO);
    if (!archivo.is_open()) {
        cout << "  No hay ventas registradas." << endl;
        return;
    }

    cout << "\n  ======== HISTORIAL DE VENTAS ========" << endl;

    string linea;
    int    contador = 0;
    double totalGeneral = 0.0;

    while (getline(archivo, linea)) {
        if (linea.empty()) continue;

        stringstream ss(linea);
        string producto, sCantidad, sPrecio, sTotal;
        getline(ss, producto,  '|');
        getline(ss, sCantidad, '|');
        getline(ss, sPrecio,   '|');
        getline(ss, sTotal,    '|');

        int    cantidad = stoi(sCantidad);
        double precio   = stod(sPrecio);
        double total    = stod(sTotal);

        Venta v(producto, cantidad, precio);
        v.mostrarFactura();

        totalGeneral += total;
        contador++;
    }

    archivo.close();

    if (contador == 0) {
        cout << " No hay ventas registradas." << endl;
    } else {
        cout << fixed << setprecision(2);
        cout << "  Total de ventas : " << contador       << endl;
        cout << "  Ingresos totales: $" << totalGeneral  << endl;
    }
}

//  Función principal – Menú interactivo
int main() {
    int opcion;

    cout << "============================================" << endl;
    cout << "   Sistema de Ventas C++ (APE05 Ej.9)    " << endl;
    cout << "============================================" << endl;

    do {
        cout << "\n  MENU PRINCIPAL" << endl;
        cout << "  1. Registrar venta"    << endl;
        cout << "  2. Historial de ventas" << endl;
        cout << "  3. Salir"              << endl;
        cout << "  Opcion: ";
        cin  >> opcion;

        switch (opcion) {
            case 1: registrarVenta();   break;
            case 2: mostrarHistorial(); break;
            case 3: cout << "  Hasta luego." << endl; break;
            default: cout << "  Opcion no valida." << endl;
        }

    } while (opcion != 3);

    return 0;
}

File: src/Java/Ejercicio9/Ej9main.java
import java.io.*;
import java.util.Scanner;

public class Ej9main {

// Archivo donde se guardan las ventas
private static final String ARCHIVO = "ventas.txt";
private static Scanner sc = new Scanner(System.in);

//  Método: registrarVenta
//  Solicita datos, genera factura y guarda en archivo
public static void registrarVenta() throws IOException {
System.out.println("\n  -- Registrar Venta --");

System.out.print("  Producto  : ");
String producto = sc.next();

int cantidad;
do {
System.out.print("  Cantidad  : ");
cantidad = sc.nextInt();
if (cantidad <= 0)
System.out.println("  La cantidad debe ser mayor a 0.");
} while (cantidad <= 0);

double precio;
do {
System.out.print("  Precio u. : ");
precio = sc.nextDouble();
if (precio <= 0)
System.out.println("  El precio debe ser mayor a 0.");
} while (precio <= 0);

Ej9Venta v = new Ej9Venta(producto, cantidad, precio);

// Mostrar factura generada
System.out.println("\n  ======== FACTURA GENERADA ========");
v.mostrarFactura();

// Guardar en archivo (modo append)
FileWriter fw   = new FileWriter(ARCHIVO, true);
BufferedWriter bw = new BufferedWriter(fw);
bw.write(v.toArchivo());
bw.newLine();
bw.close();
fw.close();

System.out.println(" Venta registrada correctamente.");
}

//  Método: mostrarHistorial
//  Lee el archivo y muestra todas las ventas
public static void mostrarHistorial() throws IOException {
File f = new File(ARCHIVO);
if (!f.exists()) {
System.out.println(" No hay ventas registradas.");
return;
}

System.out.println("\n  ======== HISTORIAL DE VENTAS ========");

BufferedReader br = new BufferedReader(new FileReader(ARCHIVO));
String linea;
int    contador     = 0;
double totalGeneral = 0.0;

while ((linea = br.readLine()) != null) {
if (linea.trim().isEmpty()) continue;
Ej9Venta v = Ej9Venta.fromArchivo(linea);
v.mostrarFactura();
totalGeneral += v.getTotal();
contador++;
}
br.close();

if (contador == 0) {
System.out.println("  [INFO] No hay ventas registradas.");
} else {
System.out.printf("  Total de ventas  : %d%n",    contador);
System.out.printf("  Ingresos totales : $%.2f%n", totalGeneral);
}
}

//  Método principal – Menú interactivo
public static void main(String[] args) {
System.out.println("============================================");
System.out.println("   Sistema de Ventas Java (APE05 Ej.9)   ");
System.out.println("============================================");

int opcion;
do {
System.out.println("\n  MENU PRINCIPAL");
System.out.println("  1. Registrar venta");
System.out.println("  2. Historial de ventas");
System.out.println("  3. Salir");
System.out.print("  Opcion: ");
opcion = sc.nextInt();

try {
switch (opcion) {
case 1: registrarVenta();   break;
case 2: mostrarHistorial(); break;
case 3: System.out.println("  Hasta luego."); break;
default: System.out.println("  Opcion no valida.");
}
} catch (IOException e) {
System.out.println("  Problema con el archivo: " + e.getMessage());
}

} while (opcion != 3);

sc.close();
}
}
